### ISSUE TITLE
add checkbox to tree

### ISSUE THREAD

**Issue by parisamashinchi** [500tech/angular-tree-component#519]
hello
I want to use  useCheckbox property  in tree options.
I use npm install --save angular-tree-component to get angular tree module,  but there is not checkbox component in it. how can i get checkbox  component ?
thanks

**Comment by Misha3739** [comment]
Hello, parisamashinchi.
Did tou find 'useCheckbox' in options object?

**Comment by parisamashinchi** [comment]
Misha 
'useCheckbox' is unknown property in options

**Comment by HexNeck** [comment]
Same problem for me. If someone could give me a hand to sort this out, please.

**Comment by isaklafleur** [comment]
`<tree-root [options]="options" [nodes]="nodes"></tree-root>`

```
  options: ITreeOptions = {
    useCheckbox: true,
    getChildren: this.getChildren.bind(this)
  };
```

Hope this helped you.

**Comment by isaklafleur** [comment]
Hi @adamkleingit 

Thank you for an amazing component! It's really cool and good documented. I have a quick question. My code below is working fine, the only this is I get an error (see below) when I select a node that has children that I have not yet expanded (I retrieve my children nodes async). I understand it tries to find the leafs (children nodes) of the parent, but cant find any so it gives this error. What can I do to prevent this?

Maybe somehow expand all nodes below if I click on a parent node that are not yet expanded?

```
ERROR TypeError: Cannot read property 'forEach' of undefined
    at TreeNode.setIsSelected (tree-node.model.js:353)
    at TreeNode.toggleSelected (tree-node.model.js:358)
    at TOGGLE_SELECTED (tree-options.model.js:6)
    at TreeNode.mouseAction (tree-node.model.js:413)
    at Object.eval [as handleEvent] (tree-node-checkbox.component.d.ts.TreeNodeCheckboxComponent.html:6)
    at handleEvent (core.js:13530)
    at callWithDebugContext (core.js:15039)
    at Object.debugHandleEvent [as handleEvent] (core.js:14626)
    at dispatchEvent (core.js:9945)
    at eval (core.js:10570)
```
```
import { Component, OnInit } from "@angular/core";
import { ITreeOptions } from "angular-tree-component";
import { EclassService } from "../../services/eclass.service";
import "rxjs/add/operator/toPromise";
import "rxjs/add/observable/from";
import "rxjs/add/operator/map";

@Component({
  selector: "app-async",
  template: `<tree-root [options]="options" [nodes]="nodes"></tree-root>`,
  styles: ["async-tree-example.component.css"]
})

export class AsyncTreeExampleComponent implements OnInit {
  options: ITreeOptions = {
    useCheckbox: true,
    getChildren: this.getChildren.bind(this)
  };
  nodes = [];
  children = [];

  constructor(private eclassService: EclassService) {}

  ngOnInit() {
    const filter: any = { level: "1" };
    this.eclassService.getEclassNodes(filter).subscribe(data => {
      this.nodes = data.map(item => {
        return {
          id: item.codedName,
          name: `${item.codedName} ${item.preferredName}`,
          hasChildren: true
        };
      });
    });
  }

  getChildren(node: any): Promise<object> {
    const filter = {
      codedName: node.id
    };

    return this.eclassService.getEclassNodes(filter)
      .map((x: any[]) => {
        return x.map(child => {
          if (child.level === "4") {
            return ({
              id: child.codedName,
              name: `${child.codedName} ${child.preferredName}`,
              hasChildren: false
            });
          } else {
            return ({
              id: child.codedName,
              name: `${child.codedName} ${child.preferredName}`,
              hasChildren: true
            });
          }
        });
      })
      .toPromise();
    }
  }

```

**Comment by phhbr** [comment]
> Hi @adamkleingit
> 
> Thank you for an amazing component! It's really cool and good documented. I have a quick question. My code below is working fine, the only this is I get an error (see below) when I select a node that has children that I have not yet expanded (I retrieve my children nodes async). I understand it tries to find the leafs (children nodes) of the parent, but cant find any so it gives this error. What can I do to prevent this?
> 
> Maybe somehow expand all nodes below if I click on a parent node that are not yet expanded?
> 
> ```
> ERROR TypeError: Cannot read property 'forEach' of undefined
>     at TreeNode.setIsSelected (tree-node.model.js:353)
>     at TreeNode.toggleSelected (tree-node.model.js:358)
>     at TOGGLE_SELECTED (tree-options.model.js:6)
>     at TreeNode.mouseAction (tree-node.model.js:413)
>     at Object.eval [as handleEvent] (tree-node-checkbox.component.d.ts.TreeNodeCheckboxComponent.html:6)
>     at handleEvent (core.js:13530)
>     at callWithDebugContext (core.js:15039)
>     at Object.debugHandleEvent [as handleEvent] (core.js:14626)
>     at dispatchEvent (core.js:9945)
>     at eval (core.js:10570)
> ```
> ```
> import { Component, OnInit } from "@angular/core";
> import { ITreeOptions } from "angular-tree-component";
> import { EclassService } from "../../services/eclass.service";
> import "rxjs/add/operator/toPromise";
> import "rxjs/add/observable/from";
> import "rxjs/add/operator/map";
> 
> @Component({
>   selector: "app-async",
>   template: `<tree-root [options]="options" [nodes]="nodes"></tree-root>`,
>   styles: ["async-tree-example.component.css"]
> })
> 
> export class AsyncTreeExampleComponent implements OnInit {
>   options: ITreeOptions = {
>     useCheckbox: true,
>     getChildren: this.getChildren.bind(this)
>   };
>   nodes = [];
>   children = [];
> 
>   constructor(private eclassService: EclassService) {}
> 
>   ngOnInit() {
>     const filter: any = { level: "1" };
>     this.eclassService.getEclassNodes(filter).subscribe(data => {
>       this.nodes = data.map(item => {
>         return {
>           id: item.codedName,
>           name: `${item.codedName} ${item.preferredName}`,
>           hasChildren: true
>         };
>       });
>     });
>   }
> 
>   getChildren(node: any): Promise<object> {
>     const filter = {
>       codedName: node.id
>     };
> 
>     return this.eclassService.getEclassNodes(filter)
>       .map((x: any[]) => {
>         return x.map(child => {
>           if (child.level === "4") {
>             return ({
>               id: child.codedName,
>               name: `${child.codedName} ${child.preferredName}`,
>               hasChildren: false
>             });
>           } else {
>             return ({
>               id: child.codedName,
>               name: `${child.codedName} ${child.preferredName}`,
>               hasChildren: true
>             });
>           }
>         });
>       })
>       .toPromise();
>     }
>   }
> ```
> image example below. If I first expand 13000000 all the way to the leaf and then click on a parent it does not give any errors, but if I would be clicking on 13010200 now I would get this error as I have not yet "downloaded" these nodes as I have not yet expanded them
> 
> ![image](https://user-images.githubusercontent.com/10033653/35590613-b3b64d9c-0607-11e8-8d54-037e52302806.png)

Could you find a solution to this issue?

**Comment by adamkleingit** [comment]
@phhbr Does your PR solve this problem as well?

**Comment by phhbr** [comment]
I cannot replicate this issue, so it should be fixed

**Comment by adamkleingit** [comment]
Published in 8.0.2 - if it doesn't solve it please reopen
